okhttp/src/main/java/okhttp3/Address.kt

```diff
@@ -0,0 +1,156 @@
+ * connections the address also includes the SSL socket factory, hostname verifier, and certificate
+ * pinner.
+ *
+ * HTTP requests that share the same `Address` may also share the same [Connection].
+ */
+class Address(
+    uriHost: String,
+    uriPort: Int,
+    private val dns: Dns,
+    private val socketFactory: SocketFactory,
+    private val sslSocketFactory: SSLSocketFactory?,
+    private val hostnameVerifier: HostnameVerifier?,
+    private val certificatePinner: CertificatePinner?,
+    private val proxyAuthenticator: Authenticator,
+    private val proxy: Proxy?,
+    protocols: List<Protocol>,
+    connectionSpecs: List<ConnectionSpec>,
+    private val proxySelector: ProxySelector
+) {
+  private val url: HttpUrl = HttpUrl.Builder()
+      .scheme(if (sslSocketFactory != null) "https" else "http")
+      .host(uriHost)
+      .port(uriPort)
+      .build()
+
+  /**
+   * The protocols the client supports. This method always returns a non-null list that
+   * contains minimally [Protocol.HTTP_1_1].
+   */
+  private val protocols: List<Protocol> = Util.immutableList(protocols)
+
+  private val connectionSpecs: List<ConnectionSpec> = Util.immutableList(connectionSpecs)
+
+  /**
+   * Returns a URL with the hostname and port of the origin server. The path, query, and fragment of
+   * this URL are always empty, since they are not significant for planning a route.
+   */
+  fun url() = url
+
+  /** Returns the service that will be used to resolve IP addresses for hostnames. */
+  fun dns() = dns
+
+  /** Returns the socket factory for new connections. */
+  fun socketFactory(): SocketFactory {
+    return socketFactory
+  }
+
+  /** Returns the client's proxy authenticator. */
+  fun proxyAuthenticator() = proxyAuthenticator
+
+  /**
+   * Returns the protocols the client supports. This method always returns a non-null list that
+   * contains minimally {@link Protocol#HTTP_1_1}.
+   */
+  fun protocols() = protocols
+
+  fun connectionSpecs() = connectionSpecs
+
+  /**
+   * Returns this address's proxy selector. Only used if the proxy is null. If none of this
+   * selector's proxies are reachable, a direct connection will be attempted.
+   */
+  fun proxySelector() = proxySelector
+
+  /**
+   * Returns this address's explicitly-specified HTTP proxy, or null to delegate to the {@linkplain
+   * #proxySelector proxy selector}.
+   */
+  fun proxy() = proxy
+
+  /** Returns the SSL socket factory, or null if this is not an HTTPS address. */
+  fun sslSocketFactory() = sslSocketFactory
+
+  /** Returns the hostname verifier, or null if this is not an HTTPS address. */
+  fun hostnameVerifier() = hostnameVerifier
+
+  /** Returns this address's certificate pinner, or null if this is not an HTTPS address. */
+  fun certificatePinner() = certificatePinner
+
+  override fun equals(other: Any?): Boolean {
+    return (other is Address
+        && url == other.url
+        && equalsNonHost((other as Address?)!!))
```

Comment:
I believe you can cast to the non-nullable type `other as Address`

Reply:
The explicit cast shouldn't even be needed. Smart cast should take care of it.

Reply:
Yup, good point! Done in e751fc152f6b74804a016a119e8fbc64ae85142f